File: LeetCode/1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.cpp
/**
 * 76 / 76 test cases passed.
 * Runtime: 376 ms
 * Memory Usage: 114.2 MB 
 */
class Solution {
public:
    int minReorder(int n, vector<vector<int>>& connections) {
        unordered_map<int, vector<int>> node2edges;
        int edges_count = connections.size();
        for (int i = 0; i < edges_count; ++ i) {
            node2edges[connections[i][0]].push_back(i);
            node2edges[connections[i][1]].push_back(i);
        }
        vector<bool> used(edges_count);
        queue<int> que;
        que.emplace(0);
        int ans = 0;
        while (!que.empty()) {
            int node = que.front(); que.pop();
            for (auto &edge_id : node2edges[node]) {
                if (used[edge_id]) continue;
                used[edge_id] = true;

                int src = connections[edge_id][0];
                int tar = connections[edge_id][1];
                if (node == src) ++ ans;
                int next = node == src ? tar : src;
                que.emplace(next);
            }
        }
        return ans;
    }
};

File: LeetCode/1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.py
"""
76 / 76 test cases passed.
Runtime: 292 ms
Memory Usage: 34.6 MB
"""
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        node2edges = collections.defaultdict(list)
        for id, edge in enumerate(connections):
            node2edges[edge[0]].append(id)
            node2edges[edge[1]].append(id)
        used = [False] * len(connections)
        que = collections.deque([0])
        ans = 0
        while que:
            node = que.pop()
            for edge_id in node2edges[node]:
                if not used[edge_id]:
                    used[edge_id] = True
                    fr = connections[edge_id][0]
                    to = connections[edge_id][1]
                    if node == fr:
                        ans += 1
                        que.append(to)
                    else:
                        que.append(fr)
        return ans
